File: prob1.py
# Time Complexity : O(n)
# Space Complexity :O(h)
# Passed on Leetcode: yes

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

import queue
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if root is None:
            return None
        q = queue.Queue()
        q.put(root)
        res = []
        while not q.empty():
            for i in range(q.qsize()):
                val = q.get()
                if val.left:
                    q.put(val.left)
                if val.right:
                    q.put(val.right)
            res.append(val.val)
        return res
            

File: prob1.java
// Time Complexity : O(N)
// Space Complexity : O(H) worsst O(N)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No

// Explain your approach: this problem can be done using level order traversal, i.e. BFS, where we keep track 
// of the size of each level and if index == size - 1, i.e last element in level, add that elem to result
// Using DFS, there are 2 ways, preorder with Root - Right - Left and when result size == level, add root.val
// or Root - Left - Right, and we overwrite the value at a particular level after the right recrusive call

import java.util.*;
public class prob1 {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    List<Integer> result = new ArrayList<>();
    public List<Integer> rightSideView(TreeNode root) {
        if(root == null) return result;
        dfs(root, 0);
        return result;
    }

    public void dfs(TreeNode root, int level){
        if(root == null) return;
        if(level == result.size()){
            result.add(root.val);  
        }
        dfs(root.left, level+1);
        dfs(root.right, level+1);
        result.set(level, root.val);
        
    }
    
}
}
